File: apps/voice/app/loop/squad.py
"""Multi-agent Squad runtime helpers (Day 27) — the Python mirror of the shared
`squad.ts` logic the live call loop consumes. A squad chains specialist agents inside one
call; the `ContextBus` carries captured state across handoffs so the next specialist never
re-asks, and `resolve_handoff` routes a turn's signal to the right specialist.

These are pure + deterministic (no I/O) so they are unit-tested without live providers,
exactly like `language.py`. The bus is created per call, so it only ever holds that one
call's data — there is no cross-tenant path."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


@dataclass(slots=True)
class HandoffRule:
    from_agent_id: str
    on: str
    to_agent_id: str


def resolve_handoff(rules: list[HandoffRule], current_agent_id: str, signal: str) -> str | None:
    """Return the next specialist's agent id for a signal, or None to keep the turn.

    First matching rule wins (declaration order)."""
    for rule in rules:
        if rule.from_agent_id == current_agent_id and rule.on == signal:
            return rule.to_agent_id
    return None


def entry_agent(members: list[dict[str, Any]], entry_agent_id: str | None) -> str | None:
    """The specialist that answers first: explicit entry, else the lowest-`order` member."""
    if entry_agent_id:
        return entry_agent_id
    if not members:
        return None
    ordered = sorted(members, key=lambda m: m.get("order", 0))
    return ordered[0].get("agentId")


@dataclass(slots=True)
class ContextBus:
    """Per-call shared state that survives handoffs. Later specialists read what earlier
    ones captured, so nothing is re-asked."""

    _store: dict[str, Any] = field(default_factory=dict)

    def set(self, key: str, value: Any, _by_agent_id: str = "system") -> None:
        if value in (None, ""):  # never store empties
            return
        self._store[key] = value

    def merge(self, record: dict[str, Any], by_agent_id: str = "system") -> None:
        for key, value in record.items():
            self.set(key, value, by_agent_id)

    def get(self, key: str) -> Any:
        return self._store.get(key)

    def has(self, key: str) -> bool:
        return key in self._store

    def snapshot(self) -> dict[str, Any]:
        return dict(self._store)

    def for_handoff(self, to_agent_id: str) -> dict[str, Any]:
        """The handoff payload seeding the next specialist: the shared snapshot plus a
        one-line summary so the receiving agent doesn't repeat questions."""
        snap = self.snapshot()
        facts = [f"{k.replace('_', ' ')}: {v}" for k, v in snap.items()]
        summary = f"Known so far — {'; '.join(facts)}." if facts else "No details captured yet."
        return {"toAgentId": to_agent_id, "context": snap, "summary": summary}


def resolve_node_override(
    override: dict[str, Any] | None, default_model: str, default_voice_id: str | None
) -> tuple[str, str | None]:
    """Resolve the effective (model, voice) for a node: node override if set, else the
    agent default. The router meters against the RESOLVED model (self-audit D)."""
    override = override or {}
    model = (override.get("modelOverride") or "").strip() or default_model
    voice = (override.get("voiceOverride") or "").strip() or default_voice_id
    return model, voice

File: packages/shared/src/squad.ts
import { z } from 'zod';

/**
 * Multi-agent Squads (Day 27). A squad chains specialist agents inside ONE live call
 * (receptionist → booking → billing). Two problems this module solves, as pure/tested
 * units the voice loop consumes:
 *
 *  1. **Handoff routing** — given the active agent + a signal (intent/keyword), which
 *     specialist takes the next turn? (`resolveHandoff`)
 *  2. **Shared context bus** — a per-call state that travels across every handoff so the
 *     next specialist never re-asks what the caller already gave. (`ContextBus`)
 *
 * The bus is created per call inside the tenant's loop, so it can only ever hold that
 * one call's data — there is no cross-tenant path (self-audit B). Per-node model/voice
 * overrides let a squad run a cheap model for routing and a premium one for high-stakes
 * turns; the router honours the resolved override + meters against it (self-audit D).
 */

// ── Squad config ──────────────────────────────────────────────────────────────

export const squadMemberSchema = z.object({
  agentId: z.string().uuid(),
  role: z.string().min(1).max(60), // receptionist / booking / billing …
  order: z.number().int().min(0).max(100).default(0),
});
export type SquadMember = z.infer<typeof squadMemberSchema>;

/**
 * A handoff rule: when `fromAgentId` is active and the turn yields signal `on`, pass the
 * call to `toAgentId`. `on` is a free label matched against the classified intent/keyword
 * ('booking', 'billing', 'human', …). First matching rule wins (declaration order).
 */
export const handoffRuleSchema = z.object({
  fromAgentId: z.string().uuid(),
  on: z.string().min(1).max(60),
  toAgentId: z.string().uuid(),
});
export type HandoffRule = z.infer<typeof handoffRuleSchema>;

export const squadConfigSchema = z
  .object({
    entryAgentId: z.string().uuid().nullish(),
    members: z.array(squadMemberSchema).max(20).default([]),
    handoffRules: z.array(handoffRuleSchema).max(100).default([]),
  })
  .superRefine((cfg, ctx) => {
    const ids = new Set(cfg.members.map((m) => m.agentId));
    // Every rule endpoint must reference a member of the squad (no dangling handoffs).
    for (const [i, r] of cfg.handoffRules.entries()) {
      if (!ids.has(r.fromAgentId)) {
        ctx.addIssue({
          code: 'custom',
          path: ['handoffRules', i, 'fromAgentId'],
          message: 'Rule references an agent not in the squad',
        });
      }
      if (!ids.has(r.toAgentId)) {
        ctx.addIssue({
          code: 'custom',
          path: ['handoffRules', i, 'toAgentId'],
          message: 'Rule references an agent not in the squad',
        });
      }
    }
    if (cfg.entryAgentId && !ids.has(cfg.entryAgentId)) {
      ctx.addIssue({
        code: 'custom',
        path: ['entryAgentId'],
        message: 'Entry agent must be a squad member',
      });
    }
  });
export type SquadConfig = z.infer<typeof squadConfigSchema>;

/** The specialist that answers first: explicit entry, else the lowest-`order` member. */
export function entryAgent(cfg: SquadConfig): string | null {
  if (cfg.entryAgentId) return cfg.entryAgentId;
  const sorted = [...cfg.members].sort((a, b) => a.order - b.order);
  return sorted[0]?.agentId ?? null;
}

/**
 * Resolve the next specialist for a signal. Returns the target agentId, or null when no
 * rule matches (the current agent keeps the turn). First matching rule wins.
 */
export function resolveHandoff(
  rules: HandoffRule[],
  currentAgentId: string,
  signal: string,
): string | null {
  const rule = rules.find((r) => r.fromAgentId === currentAgentId && r.on === signal);
  return rule?.toAgentId ?? null;
}

// ── Shared context bus ──────────────────────────────────────────────────────────

/** One fact captured during the call, tagged with the agent that captured it. */
export interface ContextEntry {
  key: string;
  value: unknown;
  byAgentId: string;
}

/**
 * Per-call shared state that survives handoffs. Later specialists read what earlier ones
 * captured, so nothing is re-asked. Instances are created per call → scoped to that call's
 * tenant by construction (never shared across calls/tenants).
 */
export class ContextBus {
  private readonly store = new Map<string, ContextEntry>();

  constructor(seed: Record<string, unknown> = {}, byAgentId = 'system') {
    for (const [key, value] of Object.entries(seed)) this.set(key, value, byAgentId);
  }

  set(key: string, value: unknown, byAgentId: string): void {
    if (value === undefined || value === null || value === '') return; // never store empties
    this.store.set(key, { key, value, byAgentId });
  }

  /** Merge a captured-variables record from a turn (last write wins per key). */
  merge(record: Record<string, unknown>, byAgentId: string): void {
    for (const [key, value] of Object.entries(record)) this.set(key, value, byAgentId);
  }

  get(key: string): unknown {
    return this.store.get(key)?.value;
  }

  has(key: string): boolean {
    return this.store.has(key);
  }

  /** Flat key→value snapshot (for templating the next agent's prompt). */
  snapshot(): Record<string, unknown> {
    const out: Record<string, unknown> = {};
    for (const [key, entry] of this.store) out[key] = entry.value;
    return out;
  }

  /**
   * The handoff payload seeding the next specialist: the shared snapshot plus a one-line
   * summary of what's already known, so the receiving agent doesn't repeat questions.
   */
  forHandoff(toAgentId: string): {
    toAgentId: string;
    context: Record<string, unknown>;
    summary: string;
  } {
    const snap = this.snapshot();
    const facts = Object.entries(snap).map(([k, v]) => `${k.replace(/_/g, ' ')}: ${String(v)}`);
    const summary = facts.length
      ? `Known so far — ${facts.join('; ')}.`
      : 'No details captured yet.';
    return { toAgentId, context: snap, summary };
  }
}

// ── Per-node model / voice override ─────────────────────────────────────────────

/**
 * Optional per-node overrides. A node may pin a specific LLM model (cheap for routing,
 * premium for high-stakes) and/or a specific voice. Empty = inherit the agent default.
 */
export const nodeOverrideSchema = z.object({
  modelOverride: z.string().max(80).default(''),
  voiceOverride: z.string().max(80).default(''),
});
export type NodeOverride = z.infer<typeof nodeOverrideSchema>;

/**
 * Resolve the effective model/voice for a node: the node override if set, else the agent
 * default. The router meters against the RESOLVED model, so a per-node swap is billed at
 * that model's rate (self-audit D).
 */
export function resolveNodeOverride(
  override: Partial<NodeOverride> | undefined,
  defaults: { model: string; voiceId: string | null },
): { model: string; voiceId: string | null } {
  return {
    model: override?.modelOverride?.trim() || defaults.model,
    voiceId: override?.voiceOverride?.trim() || defaults.voiceId,
  };
}
